### ISSUE TITLE
批处理功能，概率性异常java.lang.IllegalStateException: Cannot write to standard output when terminated

### ISSUE THREAD

**Issue by hanboyuan** [alibaba/arthas#600]
- [ ] 我已经在 [issues](https://github.com/alibaba/arthas/issues) 里搜索，没有重复的issue。

### 环境信息

* `arthas-boot.jar` 或者 `as.sh` 的版本： 3.1.0
* Arthas 版本: 3.1.0
* 操作系统版本: centos6.5
* 目标进程的JVM版本: jdk1.8.0_60
* 执行`arthas-boot`的版本: 3.1.0

### 重现问题的步骤

1. /export/servers/jdk1.8.0_60/bin/java -jar arthas-client.jar 127.0.0.1 59999 -f cpu.usage.as -w 160

cpu.usage.as内容如下
dashboard -n 1 -i 1000
thread -n 10
jvm
### 期望的结果

有时正常返回，有时在某个命令那异常，退出了

### 实际运行的结果
$ dashboard -n 1 -i 1000
ID           NAME                                    GROUP                      PRIORITY     STATE         %CPU         TIME         INTERRUPTED   DAEMON       
206          System_Clock                            main                       5            TIMED_WAITING 34           51:59        false         true         
628          SystemClock                             main                       5            TIMED_WAITING 31           55:43        false         true         
205          SystemClock                         main                       5            TIMED_WAITING 31           55:35        false         true         
8            AsyncLogger-1                           main                       5            WAITING       0            1:38         false         true         
6            AsyncLoggerConfig-1                     main                       5            WAITING       0            2:12         false         true         
82654        Attach Listener                         system                     9            RUNNABLE      0            0:0          false         true         
704          BrokenConnectionDestroyer-0-15531679064 main                       5            WAITING       0            0:3          false         true         
229          CLIENT_SIDE_RINGBUFFER                  main                       5            WAITING       0            1:26         false         true         
181          CfsHeartbeat-1553166826383              main                       5            RUNNABLE      0            0:3          false         true         
277          ChannelEvent                            main                       5            TIMED_WAITING 0            0:5          false         true         
Memory                            used        total      max         usage      GC                                                                              
heap                              1524M       2922M      2922M       52.17%     gc.parnew.count                         3244                                    
par_eden_space                    1108M       1200M      1200M       92.41%     gc.parnew.time(ms)                      1508893                                 
par_survivor_space                15M         150M       150M        10.11%     gc.concurrentmarksweep.count            1                                       
cms_old_gen                       400M        1572M      1572M       25.47%     gc.concurrentmarksweep.time(ms)         565                                     
nonheap                           222M        233M       1776M       12.52%                                                                                     
Runtime                                                                                                                                                         
os.name                                 Linux                                                                                                                   
os.version                              3.10.0-693.5.2.el7.xx_3311.x86_64                                                                                       
java.version                            1.8.0_51                                                                                                                
java.home                               /export/servers/xxk1.8.0_60/jre                                                                                         
systemload.average                      9.85                                                                                                                    
Process ends after 1 time(s).
$ thread -n 10
"SystemClock" Id=205 cpuUsage=11% TIMED_WAITING on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@44f2ddc6
    at sun.misc.Unsafe.park(Native Method)
    -  waiting on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@44f2ddc6
    at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
    at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
    at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
    at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)

"System_Clock" Id=206 cpuUsage=9% TIMED_WAITING on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@34c6dee6
    at sun.misc.Unsafe.park(Native Method)
    -  waiting on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@34c6dee6
    at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
    at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
    at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
    at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)

"SystemClock" Id=628 cpuUsage=9% TIMED_WAITING on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@3641aa4c
    at sun.misc.Unsafe.park(Native Method)
    -  waiting on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@3641aa4c
    at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
    at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
    at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
    at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)

"JSF-CLI-WORKER-5-T-4" Id=225 cpuUsage=6% RUNNABLE (in native)
    at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
    at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
    at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:79)
    at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
    -  locked io.netty.channel.nio.SelectedSelectionKeySet@6a489f4b
    -  locked java.util.Collections$UnmodifiableSet@559239d2
    -  locked sun.nio.ch.EPollSelectorImpl@9d0c3bb
    at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
    at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:622)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:310)
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112)
    at java.lang.Thread.run(Thread.java:745)

"JSF-CLI-WORKER-5-T-5" Id=226 cpuUsage=5% RUNNABLE (in native)
    at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
    at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
    at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:79)
    at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
    -  locked io.netty.channel.nio.SelectedSelectionKeySet@a652633
    -  locked java.util.Collections$UnmodifiableSet@e5ab3a7
    -  locked sun.nio.ch.EPollSelectorImpl@4b582c08
    at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
    at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:622)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:310)
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112)
    at java.lang.Thread.run(Thread.java:745)

"JSF-CLI-HB-com.xx.xxbs.price.transcenter.service.SystemSourceService-62-T-1" Id=336 cpuUsage=1% TIMED_WAITING on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@485ff1bf
    at sun.misc.Unsafe.park(Native Method)
    -  waiting on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@485ff1bf
    at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
    at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
    at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
    at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)

"JSF-CLI-HB-com.xx.xxbs.price.universal.service.PricestragyPriceSellerService-94-T-1" Id=378 cpuUsage=1% TIMED_WAITING on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@46829b9
    at sun.misc.Unsafe.park(Native Method)
    -  waiting on java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@46829b9
    at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
    at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
    at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
    at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)

jvm
"Reference Handler" Id=2 cpuUsage=0% WAITING on java.lang.ref.Reference$Lock@2c4bce66
    at java.lang.Object.wait(Native Method)
    -  waiting on java.lang.ref.Reference$Lock@2c4bce66
    at java.lang.Object.wait(Object.java:502)
    at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:157)

"Finalizer" Id=3 cpuUsage=0% WAITING on java.lang.ref.ReferenceQueue$Lock@1ad1e934
    at java.lang.Object.wait(Native Method)
    -  waiting on java.lang.ref.ReferenceQueue$Lock@1ad1e934
    at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
    at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
    at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)

Affect(row-cnt:0) cost in 934 ms.
$

```

01 2019-03-25 16:50:28.124 INFO [nioEventLoopGroup-2-1:i.n.h.l.LoggingHandler] [id: 0xd4096c8e, L:/127.0.0.1:59999] READ: [id: 0x054a8084, L:/127.0.0.1:59999 - R:/127.0.0.1:38706]
01 2019-03-25 16:50:28.125 INFO [nioEventLoopGroup-2-1:i.n.h.l.LoggingHandler] [id: 0xd4096c8e, L:/127.0.0.1:59999] READ COMPLETE
01 2019-03-25 16:50:28.161 INFO [nioEventLoopGroup-2-90:arthas] [] [] [] Loaded arthas keymap file from com/taobao/arthas/core/shell/term/readline/inputrc
01 2019-03-25 16:50:30.110 INFO [nioEventLoopGroup-2-90:arthas] [] [] [] Session cc41960c-17e4-4b05-ae49-214909c174a5 closed, so terminate foreground job, id: 264, line: jvm
01 2019-03-25 16:50:30.119 ERROR [as-command-execute-daemon:arthas] [] [] [] Error during processing the command:
java.lang.IllegalStateException: Cannot write to standard output when terminated
	at com.taobao.arthas.core.shell.system.impl.ProcessImpl$CommandProcessImpl.write(ProcessImpl.java:453) ~[arthas-core.jar:3.1.1-SNAPSHOT]
	at com.taobao.arthas.core.command.monitor200.JvmCommand.process(JvmCommand.java:81) ~[arthas-core.jar:3.1.1-SNAPSHOT]
	at com.taobao.arthas.core.shell.command.impl.AnnotatedCommandImpl.process(AnnotatedCommandImpl.java:82) ~[arthas-core.jar:3.1.1-SNAPSHOT]
	at com.taobao.arthas.core.shell.command.impl.AnnotatedCommandImpl.access$100(AnnotatedCommandImpl.java:18) ~[arthas-core.jar:3.1.1-SNAPSHOT]
	at com.taobao.arthas.core.shell.command.impl.AnnotatedCommandImpl$ProcessHandler.handle(AnnotatedCommandImpl.java:111) ~[arthas-core.jar:3.1.1-SNAPSHOT]
	at com.taobao.arthas.core.shell.command.impl.AnnotatedCommandImpl$ProcessHandler.handle(AnnotatedCommandImpl.java:108) ~[arthas-core.jar:3.1.1-SNAPSHOT]
	at com.taobao.arthas.core.shell.system.impl.ProcessImpl$CommandProcessTask.run(ProcessImpl.java:370) ~[arthas-core.jar:3.1.1-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_51]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_51]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_51]
```

**Comment by xindoo** [comment]
同遇到这个问题 
```
java.lang.IllegalStateException: Cannot write to standard output when terminated
	at com.taobao.arthas.core.shell.system.impl.ProcessImpl$CommandProcessImpl.write(ProcessImpl.java:453) ~[arthas-core.jar:3.1.0]
	at com.taobao.arthas.core.advisor.ReflectAdviceListenerAdapter.abortProcess(ReflectAdviceListenerAdapter.java:229) ~[arthas-core.jar:3.1.0]
	at com.taobao.arthas.core.command.monitor200.AbstractTraceAdviceListener.finishing(AbstractTraceAdviceListener.java:78) ~[arthas-core.jar:3.1.0]
	at com.taobao.arthas.core.command.monitor200.AbstractTraceAdviceListener.afterReturning(AbstractTraceAdviceListener.java:54) ~[arthas-core.jar:3.1.0]
	at com.taobao.arthas.core.advisor.ReflectAdviceListenerAdapter.afterReturning(ReflectAdviceListenerAdapter.java:133) ~[arthas-core.jar:3.1.0]
	at com.taobao.arthas.core.advisor.AdviceWeaver.afterReturning(AdviceWeaver.java:317) ~[arthas-core.jar:3.1.0]
	at com.taobao.arthas.core.advisor.AdviceWeaver.methodOnEnd(AdviceWeaver.java:158) ~[arthas-core.jar:3.1.0]
	at com.taobao.arthas.core.advisor.AdviceWeaver.methodOnReturnEnd(AdviceWeaver.java:111) ~[arthas-core.jar:3.1.0]
	at jdk.internal.reflect.GeneratedMethodAccessor108.invoke(Unknown Source) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
```

**Comment by lpl606606** [comment]
我在执行watch，tt等增强字节码的命令时，控制台输出cost in 98ms Connection closed by foreign host后服务进程挂掉，日志报java.lang.IllegalStateException: Cannot write to standard output when terminated，这是什么原因导致的？

**Comment by hengyunabc** [comment]
最好有可重现的案例

**Comment by lpl606606** [comment]
我的问题找到了，因为用的是IBM的jdk，可能默认不能使用agent技术，对于这种情况，应该控制台输出不支持日志且还能使用arthas命令，不应该直接就将服务挂掉，副作用太大了

**Comment by RotcodWang** [comment]
我的情况时这样，servlet RPC调用阻塞，使用arthas trace，没有看到输出。查看arthas日志，抛出异常
`java.lang.IllegalStateException: Cannot write to standard output when terminated`
当RPC阻塞结束时，arthas trace开始产生输出。

Arthas 版本: 3.1.4
操作系统版本: Red Hat Enterprise Linux Server release 7.7 
目标进程的JVM版本: openjdk1.8

**Comment by hengyunabc** [comment]
> 我的情况时这样，servlet RPC调用阻塞，使用arthas trace，没有看到输出。查看arthas日志，抛出异常
> `java.lang.IllegalStateException: Cannot write to standard output when terminated`
> 当RPC阻塞结束时，arthas trace开始产生输出。
> 
> Arthas 版本: 3.1.4
> 操作系统版本: Red Hat Enterprise Linux Server release 7.7
> 目标进程的JVM版本: openjdk1.8

可以提供可重现的demo不？

**Comment by kylixs** [comment]
fixed batch file output termination #1050

**Comment by hengyunabc** [comment]
https://github.com/alibaba/arthas/issues/1072
